src/client/dfs/dfs.c

```diff
@@ -6479,3 +6478,246 @@ dfs_obj_anchor_set(dfs_obj_t *obj, uint32_t index, daos_anchor_t *anchor)
+	}
+
+	/** descend into directories */
+	if (S_ISDIR(obj->mode))	{
+		daos_anchor_t		anchor = {0};
+		uint32_t		nr_entries = DFS_NR_ITERATE;
+
+		while (!daos_anchor_is_eof(&anchor)) {
+			rc = dfs_iterate(dfs, obj, &anchor, &nr_entries, DFS_MAX_NAME * nr_entries,
+					 oit_mark_cb, (void *)oit.cookie);
+			if (rc) {
+				D_ERROR("dfs_iterate() failed: %d\n", rc);
+				D_GOTO(out_obj, rc);
+			}
+			nr_entries = DFS_NR_ITERATE;
+		}
+	}
+
+out_obj:
+	rc = dfs_release(obj);
+	return rc;
+}
+
+int
+dfs_cont_check(daos_handle_t poh, const char *cont, uint64_t flags)
+{
+	dfs_t			*dfs;
+	daos_handle_t		coh;
+	daos_handle_t		oit;
+	daos_epoch_t		snap_epoch;
+	dfs_obj_t		*lf;
+	daos_anchor_t		anchor = {0};
+	uint32_t		nr_entries = DFS_NR_ITERATE, i;
+	daos_obj_id_t		oids[DFS_NR_ITERATE] = {0};
+	uint64_t		unmarked_entries = 0;
+	d_iov_t			marker;
+	bool			mark_data = true;
+	daos_epoch_range_t	epr;
+	int			rc, rc2;
+
+	if (flags & DFS_CHECK_LINK_LF && flags & DFS_CHECK_REMOVE) {
+		D_ERROR("can't request remove and link to l+f at the same time\n");
+		return EINVAL;
+	}
+	/** TODO - Update to use Exclusive open when available */
+	rc = daos_cont_open(poh, cont, DAOS_COO_RW, &coh, NULL, NULL);
+	if (rc) {
+		D_ERROR("daos_cont_open() failed "DF_RC"\n", DP_RC(rc));
+		return daos_der2errno(rc);
+	}
+
+	rc = dfs_mount(poh, coh, O_RDWR, &dfs);
+	if (rc) {
+		D_ERROR("dfs_mount() failed (%d)\n", rc);
+		D_GOTO(out_cont, rc);
+	}
+
+	/** create snapshot for OIT */
+	rc = daos_cont_create_snap_opt(coh, &snap_epoch, NULL, DAOS_SNAP_OPT_CR | DAOS_SNAP_OPT_OIT,
+				       NULL);
+	if (rc) {
+		D_ERROR("daos_cont_create_snap_opt failed "DF_RC"\n", DP_RC(rc));
+		D_GOTO(out_dfs, rc = daos_der2errno(rc));
+	}
+
+	/** Open OIT table */
+	rc = daos_oit_open(coh, snap_epoch, &oit, NULL);
+	if (rc) {
+		D_ERROR("daos_oit_open failed "DF_RC"\n", DP_RC(rc));
+		D_GOTO(out_snap, rc = daos_der2errno(rc));
+	}
+
+	/** get and mark the SB and root OIDs */
+	d_iov_set(&marker, &mark_data, sizeof(mark_data));
+	rc = daos_oit_mark(oit, dfs->super_oid, &marker, NULL);
+	if (rc) {
+		D_ERROR("Failed to mark SB OID in OIT: "DF_RC"\n", DP_RC(rc));
+		D_GOTO(out_oit, rc = daos_der2errno(rc));
+	}
+	rc = daos_oit_mark(oit, dfs->root.oid, &marker, NULL);
+	if (rc) {
+		D_ERROR("Failed to mark ROOT OID in OIT: "DF_RC"\n", DP_RC(rc));
+		D_GOTO(out_oit, rc = daos_der2errno(rc));
+	}
+
+	/** iterate through the namespace and mark OITs starting from the root object */
+	while (!daos_anchor_is_eof(&anchor)) {
+		rc = dfs_iterate(dfs, &dfs->root, &anchor, &nr_entries, DFS_MAX_NAME * nr_entries,
+				 oit_mark_cb, (void *)oit.cookie);
+		if (rc) {
+			D_ERROR("dfs_iterate() failed: %d\n", rc);
+			D_GOTO(out_oit, rc);
+		}
+
+		nr_entries = DFS_NR_ITERATE;
+	}
+
+	/** Create lost+found directory and properly link unmarked oids there. */
+	if (flags & DFS_CHECK_LINK_LF) {
+		rc = dfs_open(dfs, NULL, "lost+found", S_IFDIR | 0755, O_CREAT | O_RDWR, 0, 0, NULL,
```

Comment:
I expect the filename will be unique over time but there's no reason why this couldn't use a unique directory name here which might be useful if this utility is going to be run multiple times.

As this is accessible via a API then perhaps we should pass in (or out) the directory names used here.

Reply:
im not quite sure what you mean here. are you talking about the name of the directory lost+found or something else as you are referring to directories (plural)?

Reply:
This implements traditional unix semantics of putting all files into a single "lost_and_found" directory, I was wondering if creating one subdirectory per invocation and putting all recovered files from that run into the newly created subdirectory.

You'd then get "lost+found/<timestamp>/<objectid>" with a different timestamp for each time the utility is run.  Among other things it would allow users of the API iterate over newly created files without having to handle files recovered from a previous run.

Reply:
ok makes sense to me. i'll discuss this at the WG meeting. for now let's tackle this with the l+f ticket.